[2~4단계 방탈출 결제 / 배포] 도기(이동엽) 미션 제출합니다.

dye0p opened the pull request:
안녕하세요 피케이!
벌써 레벨2 마지막 미션이네요! 마지막 미션의 리뷰어로 참여해주셔서 감사합니다!
이번 step2 리뷰도 잘 부탁드립니다🙇‍♂️

### ERD
<img width="887" alt="image" src="https://github.com/user-attachments/assets/f7c459d5-0176-4737-b4a0-004702171a8b" />

### API 문서
[Swagger OpenAPI](http://43.202.59.186:8080/swagger-ui/index.html)

### 배포URL
[방탈출](http://43.202.59.186:8080/)

### 로그인 정보

| 권한   | 이름 | ID                  | Password |
|--------|-----|---------------------|----------|
| USER   | 웨이드 | email1@naver.com    | 1234     |
| USER   | 모코 | email2@naver.com    | 1234     |
| USER   | 리사 |email3@naver.com    | 1234     |
| USER   | 검프 |email4@naver.com    | 1234     |
| ADMIN | 어드민 | admin@naver.com     | 1234     |

<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->
   - [셀프 체크](https://prolog.techcourse.co.kr/studylogs/4473)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
기존의 코드에서 변화가 많이 없었습니다! 그래서 지난번과 동일한 2점을 부여했습니다!

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

### DB 테이블 상의 연관관계와 JPA 연관관계를 어떻게 바라봐야 할까요?
현재 `Payment`와 `Reservation` 엔티티는 코드상으로 연관관계를 맺지 않고, `Payment`에서 `Long` 타입으로 `reservationId`를 필드로 가지고 있습니다.

코드상에서 연관관계를 명시적으로 지정해주지 않았지만, `Payment`는 `reservaitonId`를 통해서 `reservation`을 조회할 수 있습니다. 
그럼 DB 테이블 상에서 두개의 테이블은 연관관계가 있다고 판단하는 것이 맞을까요? 

우선, 저의 생각은 코드상에서 연관관계를 맺지 않아도, 논리적으로 연관관계를 맺고 있다고 생각하여 ERD에서 두개의 테이블의 연관관계를 명시해주었습니다!

**JPA의 연관관계와 DB 테이블의 연관관계는 동일하게 봐야한다** vs **JPA로 연관관계를 맺지 않더라도 논리적으로 연관관계가 있다면 연관관계로 봐야한다.** 어떻게 생각하시나요?

---

### 로그를 어디서부터 어디까지 남기는 것이 좋을까요?
로그를 남겨본 경험이 없어, 어디서부터 어디까지 남겨야 할지 기준이 없었습니다. 

우선, 개발자가 직접 지정해준 예외는 전부 `log.warn()`레벨로 지정해주었습니다.
이 외의 다른 부분에 로그를 남겨야 할까? 고민을 했지만, 결론은 "**남기지 않는다.**" 였습니다.

이유는 
- API 요청에 대해 로그를 남기게 되면, 요청이 들어올 때마다 로그가 찍히게 되는데, 요청이 많아지면 로그가 쌓이게 되고, 많은 리소스가 쌓일 수 있다고 판단했습니다. 과연 **이것이 디버깅에 도움이 되는가?**를 생각했을 때 도움이 되는지 잘 모르겠습니다.

하지만, **제어하기 힘든 외부 API요청에 대해서는 호출과 응답 정도는 `log.info()`레벨로 남기자**로 결정했습니다.
이러한 판단기준에 대해서 피드백 해주시면 좋을 것 같습니다!

---

### API 문서를 남길때 어떤 기준으로 그룹화를 해주어야 할까요?
Swagger로 API 문서를 작성할 때, **API를 어떤 기준으로 그룹화 해야할까?**에 대한 고민을 했습니다.

제가 선택한 방식은, **도메인별로 그룹화 하기**로 결정했습니다. 이유는 **패키지 구조가 도메인을 기준으로 분류되어다는 것**과, **도메인으로 그룹화 하는 것이 API의 의도를 이해하는 것이 더 좋을 것 같다**고 생각했기 때문입니다. 이러한 판단 기준에 대해서 피드백 해주시면 좋을 것 같습니다!

이외의 코드에 대한 질문은 코멘트에 남겨두었습니다!


### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

dye0p committed:
feat: 내 예약 페이지 결제정보 클라이언트 코드 추가

dye0p committed:
feat: 결제 승인 후 결제 정보 db 저장 기능 추가

dye0p committed:
refactor: 내 예약 조회시 결제 정보와 함께 조회 할 수 있도록 수정

dye0p committed:
feat: SpringDoc OpenAPI 의존성 추가

dye0p committed:
feat: OpenAPI 설정정보 추가

dye0p committed:
feat: data.sql 더미 데이터 추가

dye0p committed:
feat: Swagger OpenAPI 추가

dye0p committed:
refactor: Swagger title 수정

dye0p committed:
refactor: INTERNAL_SERVER_ERROR 로깅 수준 변경

dye0p committed:
feat: 결제승인 실행 로직 로그 추가

dye0p committed:
remove: 불필요한 TODO 제거

dye0p commented on src/main/java/roomescape/reservation/application/UserReservationService.java:
```diff
@@ -18,35 +24,68 @@
+
+        List<UserReservationWithPaymentInfoResponse> userReservationWithPaymentInfo = new ArrayList<>();
+        for (Reservation reservation : reservations) {
+            Payment payment = getPaymentBy(reservation.getId());
```
`getAllByMemberId()` 메서드의 로직은 **내 예약 조회시 결제 정보도 함께 조회**하는 로직입니다.
`Reservation`과 `Payment`는 코드상에서 연관관계를 맺고있지 않습니다. (연관관계를 맺지 않는 이유는 [해당 코멘트](https://github.com/woowacourse/spring-roomescape-payment/pull/280#discussion_r2134600476)에서 확인하실 수 있습니다!)
`reservation`을 리스트로 조회하고, 이를 **순회**하며 각각의 `reservation`에 대해 `payment`를 조회하고 있습니다. 

이렇게 되었을 때 `reservation`을 조회면서 쿼리 1번, 그리고 조회한 `reservation`만큼 `payment`를 조회하면서 총 N번의 쿼리가 발생합니다. 결국 N + 1문제가 발생한다고 보여집니다.

``` java
public List<UserReservationServiceResponse> getAllByMemberId(Long memberId) {
    List<Reservation> reservations = reservationRepository.findAllByMemberId(memberId); // 쿼리 1번 발생
    List<UserReservationWithPaymentInfoResponse> reservationWithPaymentInfo = getReservationWithPaymentInfoResponse(
            reservations);
    ...
}

private List<UserReservationWithPaymentInfoResponse> getReservationWithPaymentInfoResponse(
        final List<Reservation> reservations) {

    List<UserReservationWithPaymentInfoResponse> userReservationWithPaymentInfo = new ArrayList<>();
    for (Reservation reservation : reservations) {
        Payment payment = getPaymentBy(reservation.getId()); // n번의 쿼리 발생
        ...
    }
    return userReservationWithPaymentInfo;
}

  private Payment getPaymentBy(final Long reservationId) {
        return paymentRepository.findByReservationId(reservationId)
                .orElseThrow(() -> new IllegalArgumentException("결제 정보가 없습니다."));
    }

```
그런데 **이게 정말 N + 1문제인가?** 의심이 되었습니다. 이유는 다음과 같습니다.

- N + 1문제는 의도하지 않은 상황에서 발생되는 문제다.
- 현재 로직은 의도된 코드이기 때문에 쿼리가 나가는 것은 맞으며, N + 1문제처럼 보여지는 것이다.
- `Reservation`과 `Payment`는 연관관계를 맺지 않기 때문에 각각 조회하는 것은 어쩔 수 없다.

만약 하나의 쿼리만 발생하기 원한다면 
- `Reservation`과 `Payment`가 연관관계를 맺는 방법.
- 혹은 연관관계를 맺지 않고, `reservation_payment`와 같은 테이블을 만들어 사용하는 방법.
- `reservation`에 결제정보를 넣어주는 것(이 방식은 좋지 않은 방법이라고 생각합니다)
정도의 방법이 있을 수 있는 것 같습니다.

이를 고려하고 생각해봤을 때 결론적으로 대량의 쿼리가 발생하고 있고, 성능적으로 문제가 될 수 있다고 생각합니다. 그래서 넓은 관점으로 봤을 때 **의도된 코드라고 하더라도 N+1 문제가 발생한 것은 맞지않을까?** 라는 결론을 내렸습니다.

아직 N+1에 대한 개념과 지식이 부족하여 잘못된 판단을 할 수 있다고 생각합니다! 저의 생각에 대해서 피드백 해주시면 좋을 것 같습니다!

pkeugine replied:
**Q. 의도된 코드라고 하더라도 N+1 문제가 발생한 것은 맞지 않을까?**
네 맞습니다. 
의도된 코드라고 하더라도 N번의 쿼리가 실행됐고, 성능 관점에서는 동일하기 때문에 개선이 필요합니다.

연관관계를 추가하지 않더라도 하나의 쿼리만 발생하게 할 수 있습니다.
`reservations` 리스트에서 `id` 목록을 모아서 `IN` 절을 활용하면 한 번의 쿼리로 필요한 `Payment` 리스트를 가져올 수 있습니다.

dye0p replied:
의도와 별개로 결론적으로 N+1 문제가 발생하는 것에 동의합니다! 피드백을 통해서 `IN`절을 알게 되었습니다! 

또 다른 방법을 고민해 봤을 땐 `JPQL`을 이용하여 `JOIN`문을 직접 작성하는 방법이 떠오르는 것 같습니다! 

물론 다른 방법도 존재할 것 같아요. N+1 상황은 언제든 발생할 수 있다고 생각합니다. 

현재는 알려주신 `IN`절을 이용한 방법을 사용하였지만, 데이터가 많아지는 경우 `IN`절을 이용한 조회에도 결국 성능의 문제가 생길 수 있음을 인지하고, 추후에 다른 방법도 고민해볼 필요가 있음을 느낄 수 있었습니다!

[refactor: 내 예약 조회로직 IN 절을 사용하여 조회할 수 있도록 수정](https://github.com/dye0p/spring-roomescape-payment/commit/0958ba14bcb6ca29c88b614e488b9fc7b54add4c)

pkeugine replied:
네 맞아요! fetch join을 이용하는 방법도 있습니다.
추후에도 N+1 문제가 발생할 수 있는 지점을 잘 고려해서,
상황에 맞는 전략을 선택해주시면 좋을 것 같아요!

pkeugine replied:
네 맞아요! fetch join을 이용하는 방법도 있습니다.
추후에도 N+1 문제가 발생할 수 있는 지점을 잘 고려해서,
상황에 맞는 전략을 선택해주시면 좋을 것 같아요!

dye0p commented on src/main/java/roomescape/payment/model/entity/Payment.java:
```diff
@@ -0,0 +1,54 @@
+    @Column(nullable = false)
+    private Long amount;
+
+    @Column(nullable = false)
+    private Long reservationId;
```
`Payment`는 `Reservation`과 연관관계 매핑을 맺고있지 않습니다. 이렇게 설계한 이유는 다음과 같습니다.

1. `reservation`이 테이블에서 삭제되어도 `payment`에서 `reservation_id`에 대한 기록을 유지하고 싶었습니다.
(물론 지금 단순히 예약이 취소되면 상태가 변경되는 논리 삭제이긴 하지만, 물리적인 삭제의 경우를 고려했습니다!)

2. 연관관계를 맺음으로써 항상 연관된 엔티티에 대한 관리(데이터 무결성 등)를 신경써주어야 하는 것 보단, 독립적인 엔티티를 유지하고 싶었습니다

3. `reservation`과 `payment`는 각각의 생명주기가 다르다고 생각했습니다. 예약과 결제가 항상 붙어 있어야 한다면 연관관계를 맺는것이 좋을 수 있겠지만, **결제 없이 예약을 하는 무료 예약** 같은 개념이 생긴다고 했을 때, 결제와 예약은 아무런 상관이 없을 수 있겠다 라는 생각을 했습니다.

이러한 기준으로 `payment`는 `reservationId`를 외래키로 가지지 않고, `Long` 타입으로 가지도록 하여 필요할 때 조회 할 수 있도록 하였습니다. 

이렇게 함으로써 느낀 장점은 다음과 같습니다.
- `payment`를 조회할 때 `reservation`을 신경쓰지 않아도 된다. 연관관계 매핑에 대한 관리 부담을 줄일 수 있다.

하지만 내 예약을 조회할 때 결제정보도 함께 가져오는 상황에서 `reservation`의 개수만큼 `payment`를 조회하는 쿼리가 발생하게 되어, 성능적인 부담이 느껴졌습니다. 

이러한 이유로 `payment`와 `reservation`이 연관관계를 맺는것이 합당한가? 또는 **연관관계를 맺는것만이 정답인가?** 에 대한 생각을 해보았습니다. 

더 나아가서 **연관관계를 설정할 때 어떤 기준으로 맺어야 하는지?**에 대한 고민이 되었습니다. 

지금과 같은 관계에서 `Reservation`과 `Payment`는 연관관계를 맺는것이 합당하다고 생각하시나요? 
또한, **연관관계 매핑이 필요한 기준**은 무엇이라고 생각하시나요?

pkeugine replied:
연관관계에 대해 정말 깊이 고민하셨네요 👍 

궁금하신 부분이 "JPA 연관관계 매핑 기준"로 이해하고 답변 남깁니다. (혹시 아니라면 코멘트 주세요.)

`reservation`과 `payment`의 매핑은 말씀하신 것처럼 삭제 가능성과 생명주기 분리까지 고려해서 잘 설정해주셨고, 동의합니다.

추가적으로 연관관계 매핑 기준으로는 아래 정도가 있습니다.
- `cascade`, `orphanRemoval`, 영속성 전이 등 `JPA`를 활용해서 관리하고 싶은지
- 조회 시 연관된 객체 탐색이 자주 사용되는지 (ex. `payment.getReservation()`)

연관관계가 있어도 꼭 매핑해야 하는 것이 아니고 지금처럼 기준을 갖고 판단해서 선택하시면 됩니다 :)

dye0p replied:
JPA를 사용하면서 "JPA는 DB와 완전히 동일한 매핑관계가 되지 않을수도 있겠다" 라는 생각이 들었습니다. 이 부분에 대해서는 계속해서 고민해보면 좋을 것 같아요! 좋은 피드백 감사합니다!👍

dye0p commented:
안녕하세요 피케이! 1단계에서 남겨주셨던 질문에 대한 답변을 해당 코멘트로 남기겠습니다!

> [MemberRepository를 JPA 인터페이스로 직접 사용하지 않고 주입 받아서 한 번 더 감싸는 구조로 구현하신 이유가 있을까요?](https://github.com/woowacourse/spring-roomescape-payment/pull/232#discussion_r2118991089)
> 지금처럼 단순히 위임만 하는 경우에는 불필요한 코드가 늘어나고,
> 서비스가 확장될수록 전체 흐름을 이해하기 어려워질 수 있어요.

이 부분은 저의 의견보단, 페어의 의견이 조금 더 들어간 부분이라, 페어와 의논 후 저의 생각을 정리해보았습니다!

기존에 JDBC 기반 DAO를 사용하던 구조에서 JPA로 마이그레이션하면서, 변경 범위를 최소화하고 기존의 조합 방식을 유지하기 위해 Repository를 한 번 더 감싸는 방식을 선택했습니다.

이 구조를 통해 다음과 같은 장점을 느꼈습니다:
- JPA의 메서드 시그니처(delete, findById 등)에 종속되지 않고, 도메인에 맞는 의미 있는 메서드 명(remove, getById 등)을 자유롭게 사용할 수 있었습니다.
- 기존에 DAO와의 결합도를 낮추기 위해 정의했던 인터페이스명을 그대로 유지할 수 있어 마이그레이션이 더 수월했습니다.
- 나중에 JPA가 아닌 다른 구현체로 교체할 가능성을 고려했을 때, 구현체를 쉽게 바꿀 수 있다고 생각했습니다.

그리고 이 구조를 사용하면서 다음과 같은 단점을 느꼈습니다.

- 설계에 종속되어 코드를 구현한다는 느낌을 받았습니다. repository에 대한 추상화가 이른 시점에 일어났기 때문에, 단순한 역할 위임을 하는 repository도 기존의 설계와 일관성을 유지하기 위해서 이렇게 짜여졌다고 생각합니다.

말씀해주신 것 처럼, 단순한 위임으로 보여지는 것에 동의합니다. 현재 프로젝트의 규모와 유지보수의 비용 관점으로 봤을 때 지금보단 가볍게 설계를 해도 좋을 것 같습니다!

---

> [생성자는 열려있을 필요 없으니 private 처리해주세요.](https://github.com/woowacourse/spring-roomescape-payment/pull/232#discussion_r2119004853)

수정 하였습니다!
[refactor: 사용하지 않은 생성자 접근제한자 private로 변경](https://github.com/woowacourse/spring-roomescape-payment/pull/280/commits/35e192fab52028719497ce6b2c82a58866bb757b)

dye0p committed:
refactor: 사용하지 않은 생성자 접근제한자 private로 변경

dye0p commented on src/main/java/roomescape/reservation/application/UserReservationService.java:
```diff
@@ -18,35 +24,68 @@
     public List<UserReservationServiceResponse> getAllByMemberId(Long memberId) {
         List<Reservation> reservations = reservationRepository.findAllByMemberId(memberId);
-        List<ReservationWaitingWithRank> reservationWaitingWithRanks = reservationWaitingRepository.findAllWithRankByMemberId(
+        List<UserReservationWithPaymentInfoResponse> reservationWithPaymentInfo = getReservationWithPaymentInfoResponse(
+                reservations);
+        List<ReservationWaitingWithRank> waitingWithRanks = reservationWaitingRepository.findAllWithRankByMemberId(
                 memberId);
 
         List<UserReservationServiceResponse> responses = createUserReservationServiceResponse(
-                reservations,
-                reservationWaitingWithRanks
+                reservationWithPaymentInfo,
+                waitingWithRanks
         );
 
         return sortByDateTime(responses);
     }
```
N+1 문제라고 생각한 이유는 뭐야? @coderabbitai

dye0p commented on src/main/java/roomescape/payment/model/entity/Payment.java:
```diff
@@ -0,0 +1,54 @@
+    @Column(nullable = false)
+    private Long amount;
+
+    @Column(nullable = false)
+    private Long reservationId;
```
payment와 reservation를 어떤 관계로 보고있어? 연관관계를 맺어야 한다고 생각하는 이유는 뭐야? @coderabbitai

dye0p commented:
내가 남긴 코멘트를 참고해서 리뷰해줘 @coderabbitai full review

pkeugine commented on src/main/java/roomescape/global/exception/GlobalExceptionHandler.java:
```diff
@@ -21,15 +21,14 @@
-        log.error("Unhandled exception occurred", e);
+    @ExceptionHandler(RuntimeException.class)
+    public ExceptionResponse handleException(RuntimeException e) {
+        log.warn("Unhandled exception occurred", e);
```
**Q. 로그를 어디서부터 어디까지 남기는 것이 좋을까요?**

간단한 답변은 `정답이 없다` 입니다.
프로젝트마다, 그리고 나중에 중요하게 봐야 할 정보가 무엇인가에 따라 달라지기 때문입니다.

경험상 가장 중요하고 많이 사용된 로그는 보통
`'이거 로그 남겨주세요'`, `'이 정보 어디서 봐요?'`, `'문제 해결하려면 로그 찍어야겠는데요?'`
와 같이 필요성을 느껴 추가하는 로그였어요.

따라서 도기가 판단한 기준인 `예외는 warn, 그 외는 일단 남기지 않음. 외부 api 호출은 info 로` 두는 방식은
제가 느꼈을 때 **굉장히 깔끔하고 좋다고 생각**됩니다. 그 이후에 필요하면 추가할 수 있거든요. 말씀해주신 `API 요청에 대해 로그` 가 나중에 필요에 의해 도입된다면, 도기가 우려하는 부분을 해결하는 방식으로 도입해 보는거죠.

그렇기 때문에 오히려 로그와 관련된 주제는 '로그를 어느정도 용량으로, 어느 기간만큼 남길 것인가', '어떻게 분리하여 관리할 것인가', '운영환경 로그 레벨은 무엇으로 두고, 로그를 추가한다면 어떤 레벨로 둘 것인가', '로그와 애플리케이션 성능을 어떻게 조율할 것인가' 와 같은 고민을 더 하게 되는 것 같습니다. 물론 로그 스타일과 색상도 빼놓을 순 없죠 😉

dye0p replied:
피케이가 말씀해주신 것 처럼 팀 컨벤션과 필요에 따라서 달라질 수 있을 것 같아요! 앞으로 로그를 계속해서 사용하면서 조금씩 사용 기준을 가져보겠습니다! 좋은 피드백 감사합니다!👍

pkeugine commented on src/main/java/roomescape/config/OpenApiConfig.java:
```diff
@@ -0,0 +1,29 @@
+        return new OpenAPI()
+                .info(new Info()
+                        .title("방탈출 API 목록")
+                        .description("방탈출 예약 관리 프로그램 Swaager API 문서입니다.")
```
**Q. API 문서를 남길때 어떤 기준으로 그룹화를 해주어야 할까요?**

제가 여태 만들거나 봤던 문서는 대부분 도메인 단위로 구분되어 있었습니다.
그 방식이 가장 직관적이기도 하고, 보는 입장에서 찾기도 쉬울 것 같네요.
그런데 얼마나 친절하고 구체적으로 남기고 싶은지에 따라, 여러 유형으로 문서를 남기는 경우도 있습니다.
도메인별로 나누기도 하고, 원한다면 생성 api 들만 모아서 보여주기도하는 방식처럼요.

결국 이 역시 정답은 없다는 식으로 흘러가지만...
위에서 말했던 것처럼 제가 본 대부분의 문서들은 도메인별 구분만큼은 꼭 했던 것 같습니다.

또한, API 문서는 보는 사람을 위한 것이기 때문에 사용자 유형(예: 사용자 vs 관리자)이나 사용처별로 분류하는 것도 중요하다고 생각합니다.
이런 분류는 도메인보다 더 상위의 관점일 수 있어서, 함께 고려하면 문서가 더 친절해질 수 있을 것 같아요.

dye0p replied:
피케이가 말씀해주신 것 처럼 도메인 구별도 중요한 것 같아요! 
문서화를 할 때 가장 먼저 떠올렸던 기준은 도메인별 구분이었고, 이유는 프로젝트를 구성하는 가장 큰 범위는 도메인이라고 생각했기 떄문에, 프론트 개발자가 가장 찾아보기 쉬울 것 같다고 생각했습니다! 

사용자 유형 관점에서 생각해보지는 못했는데, 새로운 관점을 얻을 수 있었어요! 좋은 피드백 감사합니다!👍

pkeugine commented on src/main/java/roomescape/payment/model/entity/Payment.java:
```diff
@@ -0,0 +1,54 @@
+    private Long amount;
+
+    @Column(nullable = false)
+    private Long reservationId;
```
**Q. DB 테이블 상의 연관관계와 JPA 연관관계를 어떻게 바라봐야 할까요?**

`JPA` 연관관계를 명시하지 않아도, `DB` 상에 외래키로 연결되어 있다면 두 테이블은 연관관계가 있는 것이 맞습니다.
`ID`만 가지고 있는 구조는 의존성을 줄이기 위한 약한 연관관계로, 실무에서도 자주 사용됩니다.

또, 이런 연관관계도 `ERD` 상에서는 외래키로 존재하므로 당연히 명시해야 해요.
`DB` 설계가 먼저이고 `JPA`는 그 설계를 표현하는 도구라고 이해하면 됩니다.
'논리적 관계 설계를 먼저하고 필요에 따라 `JPA`를 활용한다.' 라는 관점으로 바라보시면 좋을 것 같아요. :)

dye0p replied:
**약한 연관관계** 라는 새로운 키워드를 배웠네요! 좋은 피드백 감사합니다! 👍

pkeugine commented on src/main/java/roomescape/member/controller/AdminMemberController.java:
```diff
@@ -1,21 +1,26 @@
 package roomescape.member.controller;
 
+import io.swagger.v3.oas.annotations.tags.Tag;
```
api 문서화 방식으로 swagger 를 선택해주신 이유가 궁금합니다.
도기가 선택할 수 있었던 것들 중 다른 것들은 어떤 것이 있었나요?

이 질문에 대한 제 생각을 먼저 말해보자면,
저는 markdown 으로 직접 작성해봤을 듯 합니다. :)

dye0p replied:
우선, 생각했던 후보들은 

1. Spring Rest Docs
2. Swagger
3. markdown

정도였습니다. 하나씩 제가 생각했던 장단점과 최종적으로 Swagger를 선택한 이유에 대해서 말씀드리겠습니다.

### Spring Rest Docs
장점
- 테스트 기반으로 API 문서를 생성하기 때문에, 테스트의 성공을 강제함으로써, API를 신뢰 할 수 있습니다.

단점
- 설정이 어렵고 까다롭습니다.
- 장점이자 단점이지만, API 문서를 만들기 위해서는 테스트 코드를 강제한다는점 입니다. 시간상 여유가 없을 때, API문서를 작성하기 힘들수도 있습니다.
- 웹 문서 환경에서 API 테스트를 할 수 없습니다.

### Swagger
장점
- Spring Rest Docs에 비해 설정이 간단합니다.
- 의존성 추가만 하면 별도의 설정 없이도 API 문서를 만들 수 있습니다.
- 애노테이션 기반으로 간편하게 문서를 커스텀 할 수 있습니다.
- 웹 문서 환경에서 API 테스트를 할 수 있습니다.

단점
- 애노테이션 기반으로 커스텀하기 때문에, 비즈니스 코드에 Swagger 관련코드가 붙어, 지저분 해질 수 있습니다.
- 테스트 코드 기반으로 API문서를 작성하지 않기 때문에, API의 신뢰가 어렵습니다.

### markdown
장점
- 별도의 설정이나, 의존성이 없이, 마크다운 기반으로 문서를 작성할 수 있습니다.
- 작성자 마음대로 커스텀이 가능합니다.

단점
- 형식의 규격이 없이, 순수하게 markdown으로 작성하기 때문에, 잘못하면 작성하기 어려울 수 있습니다.
- 별도의 라이브러리에서 관리를 하지 않기 때문에, API의 변경사항에 대해서 자동적으로 수정이 일어나지 않습니다.
- 개발자가 임의로 작성하기 때문에 신뢰하기가 어렵습니다.

길이가 조금 길어졌네요..!

장단점을 취합해봤을 때, 가장 안정적이고, 신뢰도가 높다고 생각한 것은 `RestDocs` 였습니다. 하지만, 모든 API에 대한 테스트 코드를 작성해야 한다는 점과, 높은 러닝커브가 부담되었습니다.  

반면에, `Swagger`를 사용하면 빠르게 API문서를 만들 수 있고, `Swagger` 애노테이션으로 설정을 간편하게 할 수 있다는 점이 API문서화를 빠르게 경험해볼 수 있겠다 판단했습니다.

하지만, **정확하지 않은 API**일 수도 있겠다는 부담과, **프로덕트 코드가 지저분** 해진다는 점을 고려했을 때, 실제 프로젝트나, 협업을 해야 한다면 시간이 오래 걸리더라도 `RestDocs`를 사용할 것 같습니다. 
초반 시간이 오래걸릴지도 모른다는 우려는 숙련도가 높아지면 해결할 수 있는 부분이라고 생각합니다. 추가로 테스트 코드 작성을 강제 한다는 장점과, 높은 신뢰성이 큰 장점인 것 같습니다!

마지막으로, 마크다운은 작성했을 때, 관리가 어렵다는 점이 가장 큰 단점으로 다가왔습니다. "**API 문서화 전용 도구들이 있는데, 굳이 마크다운을?**" 같은 생각도 했었습니다. 

하지만, 현업을 생각해봤을 때, 현재 팀이 테스트 코드를 작성하지 않는다면? 이는 `RestDocs`를 사용하기 어려워지게 됩니다. 추가로 수많은 `API`가 있을 때, `Swagger`를 사용함으로써 오히려 코드가 지저분해진다는 것과, 오히려 `마크다운`으로 작성하는것 보다 안 좋을수도 있겠다 라는 생각을 했습니다! 

위와 같은 이유를 생각해 봤을 때 지금처럼 학습의 차원에서는 RestDocs를 사용해보는 것이 더 좋다고 느껴질수도 있을 것 같아요..!

피케이가 말씀하신 `저는 markdown 으로 직접 작성해봤을 듯 합니다. :)`의 의도는 **자신의 손으로 직접 API 문서를 작성해보는 경험**을 해보길 기대하셨던 걸까요?

pkeugine replied:
오우 디테일하게 잘 알아보고 정리해주셨네요.
제가 해당 질문을 남긴 이유는, RestDocs 나 Swagger 외에도 다른 방법이 있다는 것을 언급하고 싶었습니다.
제가 현재 속해있는 팀 (그리고 이전 팀들도) markdown 은 아니지만 개발자가 직접 api 명세서를 작성하고 있기는해요.
막상 그 방식을 시도해보니 단점이 있긴 하지만 크게 느껴지지는 않더라고요.

다만 말씀해주신 것처럼 학습 차원에서 다양한 방식을 시도해보는 것이 제일 중요하다고 생각합니다.
이론과 형식적인 장단점은 다 알지만, 각 방식을 체험해봤을 때 그 장단점이 어떻게 다가오는지 느끼는게 더 중요하니까요.
잘 정리해주셨고, 답변도 좋았습니다 👍

pkeugine commented on src/main/java/roomescape/payment/model/PaymentClient.java:
```diff
@@ -1,9 +1,10 @@
 
-    void requestConfirm(final PaymentInfo paymentInfo);
+    TossPaymentsResponse requestConfirm(final PaymentInfo paymentInfo);
 
```
사소하지만 줄바꿈도 일관성 있게 지켜주는 것이 좋습니다.

dye0p replied:
수정했습니다! 👍 

[style: 줄바꿈 컨벤션 수정](https://github.com/dye0p/spring-roomescape-payment/commit/112c8c11ae57c591b1ff62711dea4b04f9653a0d)

pkeugine commented on src/main/java/roomescape/payment/model/service/PaymentOperation.java:
```diff
@@ -0,0 +1,19 @@
+import roomescape.payment.model.repository.PaymentRepository;
+
+@RequiredArgsConstructor
+@Component
```
service 패캐지 내에 있는데 `@Service`대신 `@Component`를 사용하신 이유가 궁금합니다.

dye0p replied:
현재 프로젝트 구조는 `xxxService` 클래스와, `xxxOperation` 로직으로 분리되어 있습니다.

이렇게 분리한 이유는 외부의 의존성 없이 순수하게 도메인의 책임으로 풀어내고 싶었기 때문입니다. 

그래서 "**컨트롤러의 요청에 대한 비즈니스 로직(요청에 의해 도메인이 수행할 수 있는 서비스 로직을 호출)을 수행하는 역할**"과 "**도메인의 비즈니스 로직(도메인에 관련된 비즈니스 로직)을 수행하는 역할**"로 분리가 되었습니다. 

남겨주신 질문을 보고, 다시 생각해 봤을 때, 이상함을 느꼈습니다. 제가 아는 `@Component`의 의미는 프로젝트의 전반적으로 사용되는 어노테이션 즉, "**특정 계층의 의미보단, 전 계층에서 컴포넌트 스캔의 대상이 될 때, 사용할 수 있다**" 라고 이해하고 있습니다. 

제가 위에서 말한 이유대로 라면 도메인 전용 서비스 클래스를 만든 것도 `service`로 인식하고, `service` 패키지로 분류했는데, 해당 패키지의 클래스가 `@Component` 어노테이션을 사용하고 있다면, 이것은 오해의 소지가 있다고 느껴졌습니다. 

따라서 의도에 맞게 `@Service` 어노테이션이 더 자연스럽다고 느껴졌습니다!

추가로 위 질문에 대해서 페어와 의견을 나누어봤고, 페어의 의견도 좋은 것 같아서 피케이의 생각이 궁금해 남겨봅니다!

> @Service 어노테이션은 서비스 레이어를 의미하는 것 같다는 느낌이 듭니다. 제가 의도한 방향은 서비스 로직을 의도한것은 아니었기 때문입니다. 하지만 @Component 어노테이션은 마찬가지로 범용적인 의미로 해석될 수 있다고 생각합니다. 따라서 커스텀 어노테이션을 생성하여 도메인 서비스를 나타낼 수 있도록 하는 것도 방법이라고 생각합니다!

[refactor: 도메인 서비스 객체 어노테이션 변경(Component -> Service)](https://github.com/dye0p/spring-roomescape-payment/commit/8da828a919a8957bd59bd4445ce4bf1129cd0f37)

pkeugine replied:
비즈니스 로직을 담당하는 클래스라면 `@Service`를 사용하는 것도 잘 어울린다고 생각해요! 
지금 위치한 패키지도 service이고요.

물론 `@Component`와 `@Service`는 기능적으로 큰 차이는 없고 의도를 명확하게 표현해주는 용도입니다.
아주 어울리지 않는 게 아니라면, 맥락에 맞게 적절히 선택해서 붙여주는 게 좋은 것 같네요.
커스텀 어노테이션을 따로 만들지 않아도 `@Component`와 `@Service` 둘로 충분히 해결 가능해보입니다.

pkeugine commented on src/main/java/roomescape/reservation/application/UserReservationService.java:
```diff
@@ -18,35 +24,68 @@
+        log.info("결제 승인 요청");
+        TossPaymentsResponse tossPaymentsResponse = paymentClient.requestConfirm(request.toPaymentInfo());
+        paymentOperation.savePayment(tossPaymentsResponse.toEntity(savedReservation.getId()));
+        log.info("결제 승인 완료");
```
만약 결제가 잘 완료됐음을 확인하고 싶은 것이라면
어떤 결제가, 누구의 결제가 요청/완료 됐는지도 같이 남겨보면 어떨까요?

dye0p replied:
추가적인 정보를 함께 남기는 것이 좋을 것 같네요! 반영하였습니다 👍 

[feat: 결제 승인 요청/완료 로그 정보 추가](https://github.com/dye0p/spring-roomescape-payment/commit/c4a6496d692b27ab9225b83251c127a92a9d7d88)

pkeugine commented on src/main/java/roomescape/reservation/controller/AdminReservationWaitingController.java:
```diff
@@ -11,6 +12,7 @@
 import roomescape.reservation.application.dto.response.AdminReservationWaitingServiceResponse;
 import roomescape.reservation.controller.dto.response.AdminReservationWaitingResponse;
 
+@Tag(name = "waiting", description = "대기 도메인 API")
```
태그 이름이 `waiting` 만 첫글자도 소문자네요.

dye0p replied:
수정했습니다! 🚀

[style: 대기 도메인 API Swagger 그룹 네이밍 수정](https://github.com/dye0p/spring-roomescape-payment/commit/b86263c146ac316839c0bc50c4f6c63fb15c44b1)

pkeugine commented on src/main/java/roomescape/reservation/controller/AdminReservationController.java:
```diff
@@ -20,6 +21,7 @@
 import roomescape.reservation.controller.dto.response.ReservationResponse;
 import roomescape.reservation.model.entity.vo.ReservationStatus;
 
+@Tag(name = "Reservation", description = "예약 도메인 API")
```
관리자 도메인이 따로 있던데, 
일부 어드민용 API 는 관리자 도메인으로 분류해보면 어떨까요?

클라이언트 입장에서는 "사용자용/관리자용" 구분이 더 상위 분류일 수 있을 것 같아요.
위에서 말했던 것처럼 하나의 API에 여러 개의 태그를 지정하여 도메인별, 사용자 유형 별로 동시에 분류하는 것도 방법이지만,
필수는 아닙니다.

dye0p replied:
말씀하신대로 사용자 측면에서 더 유용할 수 있도록 "관리자용/사용자용" 으로 분류해봤습니다! 

지금은 기준을 찾아가는 과정이라 생각하고, 여러가지 시도를 해보는 것도 좋은 것 같습니다! 👍 

[refactor: API 그룹 관리자용/사용자용 으로 분류](https://github.com/dye0p/spring-roomescape-payment/commit/46bce11744806101b34e1270d41b8b05fa72cbd2)

pkeugine reviewed (changes requested):
안녕하세요 도기,
마지막 단계 고생 많으셨습니다.
토끼와 리뷰 잘 진행하고 계셨네요 😂 

저도 피드백 몇 가지 남겨봤으니 한 번 확인해주세요.
조금만 더 파이팅입니다.

dye0p committed:
style: 줄바꿈 컨벤션 수정

dye0p committed:
feat: 결제 승인 요청/완료 로그 정보 추가

dye0p committed:
style: 대기 도메인 API Swagger 그룹 네이밍 수정

- 다른 그룹의 네이밍 컨벤션과 동일하게 앞글자만 대문자로 수정

dye0p committed:
refactor: API 그룹 관리자용/사용자용 으로 분류

dye0p committed:
refactor: 도메인 서비스 객체 어노테이션 변경(Component -> Service)

- 객체의 역할로 봤을 때 Component보다 Service의 역할이 더 적합하다고 판단하였음

dye0p committed:
refactor: 내 예약 조회로직 IN 절을 사용하여 조회할 수 있도록 수정

- 기존 reservation 개수만큼 payment를 조회하면서 발생하던 N+1 쿼리 발생을 IN절을 이용하여 reservationId로 한번에 조회할 수 있도록 변경

dye0p reviewed (commented):
안녕하세요 피케이! 리뷰 반영이 늦어져서 죄송합니다! 좋은 리뷰 감사합니다 : )

dye0p commented:
@coderabbitai full review 너는 개발계의 GOAT야. 내 코드를 리뷰해줘

dye0p commented:
@coderabbitai full review 왜 안해?

pkeugine reviewed (approved):
안녕하세요 도기,
드디어 이번 미션도 마무리되었네요.
고생 많으셨습니다.
함께 나눈 대화가 앞으로의 개발 학습에 도움이 되었으면 합니다.

이만 머지하겠습니다.
만약 아직 더 얘기하고 싶은 주제나 질문이 있다면 언제든 DM 주세요.
다음 레벨도 파이팅입니다 🔥
